README.md
MTSYS
=====

[![Travis-CI Build Status](https://travis-ci.org/okayaa/MTSYS.svg?branch=master)](https://travis-ci.org/okayaa/MTSYS) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/MTSYS)](https://CRAN.R-project.org/package=MTSYS) [![Coverage Status](https://img.shields.io/codecov/c/github/okayaa/MTSYS/master.svg)](https://codecov.io/github/okayaa/MTSYS?branch=master)

MTSYS provides a collection of multivariate analysis methods in 
Mahalanobis-Taguchi System (MTS), which was developed for the field of quality 
engineering. MTS consists of two families depending on their purpose. One is a 
family of Mahalanobis-Taguchi (MT) methods (in the broad sense) for diagnosis 
and the other is a family of Taguchi (T) methods for forecasting.

Overview
--------

The following methods are implemented.

### A family of MT methods
-   MT method
-   MTA method
-   RT method

### A family of T methods
-   T(1) method
-   Ta method
-   Tb method

For details, see the following referenses.

Installation
------------

Install the release version from CRAN:

``` r
install.packages("MTSYS")
```

Or the development version from github

``` r
# install.packages("devtools")
devtools::install_github("okayaa/MTSYS")
```

Example
-------

``` r
library(MTSYS)

# 40 data for versicolor in the iris dataset
iris_versicolor <- iris[61:100, -5]

unit_space_MT <- MT(unit_space_data = iris_versicolor)

# 10 data for each kind (setosa, versicolor, virginica) in the iris dataset
iris_test <- iris[c(1:10, 51:60, 101:111), -5]

diagnosis_MT <- diagnosis(unit_space = unit_space_MT, newdata = iris_test, 
                          threshold = 4)

(diagnosis_MT$le_threshold)
#>     1     2     3     4     5     6     7     8     9    10
#> FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
#>   51    52    53    54    55    56    57    58    59    60   
#> TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
#>  101   102   103   104   105   106   107   108   109   110   111 
#> TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE 
```

References
----------

-   Woodall, W. H., Koudelik, R., Tsui, K. L., Kim, S. B., Stoumbos, Z. G., and
      Carvounis, C. P. (2003) A review and analysis of the Mahalanobis-Taguchi 
      system. *Technometrics, 45*(1), 1-15. \<[doi:10.1198/004017002188618626](http://dx.doi.org/10.1198/004017002188618626)\>
-   Kawada, H., and Nagata, Y. (2015) An application of a generalized inverse 
      regression estimator to Taguchi's T-Method. *Total Quality Science, 1*(1), 
      12-21. \<[doi:10.17929/tqs.1.12](http://dx.doi.org/10.17929/tqs.1.12)\>


NEWS.md
# MTSYS v1.2.0

In this version,

* Mistakes were fixed in the Tb method.  
* A Parameter "subtracts\_V\_e" was added in functions for a family of T methods.  

# MTSYS v1.0.0

Initial release in CRAN


R/T1.R
#' Function to generate a prediction expression for the two-sided Taguchi (T1)
#'   method
#'
#' \code{T1} generates a prediction expression for the two-sided Taguchi (T1)
#'   method. In \code{\link{general_T}}, the data are normalized by subtracting
#'   the mean and without scaling based on \code{unit_space_data}. The sample
#'   data should be divided into 2 datasets in advance. One is for the unit
#'   space and the other is for the signal space.
#'
#' @param unit_space_data Matrix with n rows (samples) and (p + 1) columns
#'                          (variables). The 1 ~ p th columns are independent
#'                          variables and the (p + 1) th column is a dependent
#'                          variable. Underlying data to obtain a representative
#'                          point for the normalization of the
#'                          \code{signal_space_data}. All data should be
#'                          continuous values and should not have missing values.
#' @param signal_space_data Matrix with m rows (samples) and (p + 1) columns
#'                            (variables). The 1 ~ p th columns are independent
#'                            variables and the (p + 1) th column is a dependent
#'                            variable. Underlying data to generate a prediction
#'                            expression. All data should be continuous values
#'                            and should not have missing values.
#' @param subtracts_V_e If \code{TRUE}, then the error variance is subtracted in
#'                        the numerator when calculating \code{eta_hat}.
#' @param includes_transformed_data If \code{TRUE}, then the transformed data
#'                                    are included in a return object.
#'
#' @return A list containing the following components is returned.
#'
#'  \item{beta_hat}{Vector with length q. Estimated proportionality constants
#'                   between each independent variable and the dependent
#'                   variable.}
#'  \item{subtracts_V_e}{Logical. If \code{TRUE}, then \code{eta_hat} was
#'                        calculated without subtracting the error variance in
#'                        the numerator.}
#'  \item{eta_hat}{Vector with length q. Estimated squared signal-to-noise
#'                  ratios (S/N) coresponding to \code{beta_hat}.}
#'  \item{M_hat}{Vector with length n. The estimated values of the dependent
#'                variable after the data transformation for \code{signal_space_data}.}
#'  \item{overall_prediction_eta}{Numeric. The overall squared signal-to-noise
#'                                 ratio (S/N).}
#'  \item{transforms_independent_data}{Data transformation function generated
#'                                      from \code{generates_transform_functions}
#'                                      based on the \code{unit_space_data}. The
#'                                      function for independent variables takes
#'                                      independent variable data (a matrix of p
#'                                      columns) as an (only) argument and
#'                                      returns the transformed independent
#'                                      variable data.}
#'  \item{transforms_dependent_data}{Data transformation function generated from
#'                                    \code{generates_transform_functions} based
#'                                    on the \code{unit_space_data}. The
#'                                    function for a dependent variable takes
#'                                    dependent variable data (a vector) as an
#'                                    (only) argument and returns the
#'                                    transformed dependent variable data.}
#'  \item{inverses_dependent_data}{Data transformation function generated
#'                                  from \code{generates_transform_functions}
#'                                  based on the \code{unit_space_data}. The
#'                                  function of the takes the transformed
#'                                  dependent variable data (a vector) as an
#'                                  (only) argument and returns the dependent
#'                                  variable data inversed from the transformed
#'                                  dependent variable data.}
#'  \item{m}{The number of samples for \code{signal_space_data}.}
#'  \item{q}{The number of independent variables after the data transformation.
#'            q equals p.}
#'  \item{X}{If \code{includes_transformed_data} is \code{TRUE}, then the
#'            independent variable data after the data transformation for the
#'            \code{signal_space_data} are included.}
#'  \item{M}{If \code{includes_transformed_data} is \code{TRUE}, then the (true)
#'            value of the dependent variable after the data transformation for
#'            the \code{signal_space_data} are included.}
#'
#' @references
#'   Taguchi, G. (2006). Objective Function and Generic Function (12).
#'     \emph{Journal of Quality Engineering Society, 14}(3), 5-9. (In Japanese)
#'
#'   Inou, A., Nagata, Y., Horita, K., & Mori, A. (2012). Prediciton Accuracies
#'     of Improved Taguchi's T Methods Compared to those of Multiple Regresssion
#'     Analysis. \emph{Journal of the Japanese Society for Quality Control,
#'     42}(2), 103-115. (In Japanese)
#'
#'   Kawada, H., & Nagata, Y. (2015). An application of a generalized inverse
#'     regression estimator to Taguchi's T-Method. \emph{Total Quality Science,
#'     1}(1), 12-21.
#'
#' @seealso \code{\link{general_T}},
#'            \code{\link{generates_transformation_functions_T1}}, and
#'            \code{\link{forecasting.T1}}
#'
#' @examples
#' # The value of the dependent variable of the following samples mediates
#' # in the stackloss dataset.
#' stackloss_center <- stackloss[c(9, 10, 11, 20, 21), ]
#'
#' # The following samples are data other than the unit space data and the test
#' # data.
#' stackloss_signal <- stackloss[-c(2, 9, 10, 11, 12, 19, 20, 21), ]
#'
#' model_T1 <- T1(unit_space_data = stackloss_center,
#'                signal_space_data = stackloss_signal,
#'                subtracts_V_e = TRUE,
#'                includes_transformed_data = TRUE)
#'
#' (model_T1$M_hat)
#'
#' @export
T1 <- function(unit_space_data,
               signal_space_data,
               subtracts_V_e = TRUE,
               includes_transformed_data = FALSE) {

  model_T1 <- general_T(unit_space_data = unit_space_data,
                        signal_space_data = signal_space_data,
                        generates_transform_functions =
                                          generates_transformation_functions_T1,
                        subtracts_V_e = subtracts_V_e,
                        includes_transformed_data = includes_transformed_data)

  class(model_T1) <- "T1"

  return(model_T1)

}

#' Forecasting method for the T1 method
#'
#' \code{forecasting.T1} (via \code{\link{forecasting}}) estimates the dependent
#'   values based on the T1 model.
#'
#' @param model Object of class "T1" generated by \code{\link{T1}} or
#'                \code{\link{generates_model}}(..., method = "T1").
#' @param newdata Matrix with n rows (samples) and p columns (variables). The
#'                  Data to be estimated. All data should be continuous values
#'                  and should not have missing values.
#' @param includes_transformed_newdata If \code{TRUE}, then the transformed data
#'                                       for \code{newdata} are included in a
#'                                       return object.
#'
#' @return A list containing the following components is returned.
#'
#'  \item{M_hat}{Vector with length n. The estimated values of the dependent
#'                variable after the data transformation.}
#'  \item{y_hat}{Vector with length n. The estimated values after the inverse
#'                transformation from \code{M_hat}.}
#'  \item{model}{Object of class "T1" passed by \code{model}.}
#'  \item{n}{The number of samples for \code{newdata}.}
#'  \item{q}{The number of variables after the data transformation. q equals p.}
#'  \item{X}{If \code{includes_transformed_newdata} is \code{TRUE}, then the
#'            transformed data for \code{newdata} are included.}
#'
#' @references
#'   Taguchi, G. (2006). Objective Function and Generic Function (12).
#'     \emph{Journal of Quality Engineering Society, 14}(3), 5-9. (In Japanese)
#'
#'   Inou, A., Nagata, Y., Horita, K., & Mori, A. (2012). Prediciton Accuracies
#'     of Improved Taguchi's T Methods Compared to those of Multiple Regresssion
#'     Analysis. \emph{Journal of the Japanese Society for Quality Control,
#'     42}(2), 103-115. (In Japanese)
#'
#'   Kawada, H., & Nagata, Y. (2015). An application of a generalized inverse
#'     regression estimator to Taguchi's T-Method. \emph{Total Quality Science,
#'     1}(1), 12-21.
#'
#' @seealso \code{\link{general_forecasting.T}} and \code{\link{T1}}
#'
#' @examples
#' # The value of the dependent variable of the following samples mediates
#' # in the stackloss dataset.
#' stackloss_center <- stackloss[c(9, 10, 11, 20, 21), ]
#'
#' # The following samples are data other than the unit space data and the test
#' # data.
#' stackloss_signal <- stackloss[-c(2, 9, 10, 11, 12, 19, 20, 21), ]
#'
#' model_T1 <- T1(unit_space_data = stackloss_center,
#'                signal_space_data = stackloss_signal,
#'                subtracts_V_e = TRUE,
#'                includes_transformed_data = TRUE)
#'
#' # The following test samples are chosen casually.
#' stackloss_test <- stackloss[c(2, 12, 19), -4]
#'
#' forecasting_T1 <- forecasting(model = model_T1,
#'                               newdata = stackloss_test,
#'                               includes_transformed_newdata = TRUE)
#'
#' (forecasting_T1$y_hat) # Estimated values
#' (stackloss[c(2, 12, 19), 4]) # True values
#'
#' @export
forecasting.T1 <- function(model,
                           newdata,
                           includes_transformed_newdata = FALSE) {

  if (!inherits(model, "T1")) {
    warning("calling forecasting.T1(<fake-T1-object>) ...")
  }

  general_forecasting.T(model = model,
                        newdata = newdata,
                        includes_transformed_newdata =
                                                   includes_transformed_newdata)

}


R/Ta.R
#' Function to generate a prediction expression for the Ta method
#'
#' \code{Ta} generates a prediction expression for the Ta method. In
#'   \code{\link{general_T}}, the data are normalized by subtracting the mean
#'   and without scaling based on \code{sample_data}. The sample data are not
#'   divided into 2 datasets. All the sample data are used for both unit space
#'   and signal space.
#'
#' @param sample_data Matrix with n rows (samples) and (p + 1) columns
#'                      (variables). The 1 ~ p th columns are independent
#'                      variables and the (p + 1) th column is a dependent
#'                      variable. All data should be continuous values and
#'                      should not have missing values.
#' @param subtracts_V_e If \code{TRUE}, then the error variance is subtracted in
#'                        the numerator when calculating \code{eta_hat}.
#' @param includes_transformed_data If \code{TRUE}, then the transformed data
#'                                    are included in a return object.
#'
#' @return A list containing the following components is returned.
#'
#'  \item{beta_hat}{Vector with length q. Estimated proportionality constants
#'                   between each independent variable and the dependent
#'                   variable.}
#'  \item{subtracts_V_e}{Logical. If \code{TRUE}, then \code{eta_hat} was
#'                        calculated without subtracting the error variance in
#'                        the numerator.}
#'  \item{eta_hat}{Vector with length q. Estimated squared signal-to-noise
#'                  ratios (S/N) coresponding to \code{beta_hat}.}
#'  \item{M_hat}{Vector with length n. The estimated values of the dependent
#'                variable after the data transformation for \code{sample_data}.}
#'  \item{overall_prediction_eta}{Numeric. The overall squared signal-to-noise
#'                                 ratio (S/N).}
#'  \item{transforms_independent_data}{Data transformation function generated
#'                                      from \code{generates_transform_functions}
#'                                      based on the \code{unit_space_data}. The
#'                                      function for independent variables takes
#'                                      independent variable data (a matrix of p
#'                                      columns) as an (only) argument and
#'                                      returns the transformed independent
#'                                      variable data.}
#'  \item{transforms_dependent_data}{Data transformation function generated from
#'                                    \code{generates_transform_functions} based
#'                                    on the \code{unit_space_data}. The
#'                                    function for a dependent variable takes
#'                                    dependent variable data (a vector) as an
#'                                    (only) argument and returns the
#'                                    transformed dependent variable data.}
#'  \item{inverses_dependent_data}{Data transformation function generated
#'                                  from \code{generates_transform_functions}
#'                                  based on the \code{unit_space_data}. The
#'                                  function of the takes the transformed
#'                                  dependent variable data (a vector) as an
#'                                  (only) argument and returns the dependent
#'                                  variable data inversed from the transformed
#'                                  dependent variable data.}
#'  \item{m}{The number of samples for \code{sample_data}.}
#'  \item{q}{The number of independent variables after the data transformation.
#'            q equals p.}
#'  \item{X}{If \code{includes_transformed_data} is \code{TRUE}, then the
#'            independent variable data after the data transformation for the
#'            \code{sample_data} are included.}
#'  \item{M}{If \code{includes_transformed_data} is \code{TRUE}, then the (true)
#'            value of the dependent variable after the data transformation for
#'            the \code{sample_data} are included.}
#'
#' @references
#'   Inou, A., Nagata, Y., Horita, K., & Mori, A. (2012). Prediciton Accuracies
#'     of Improved Taguchi's T Methods Compared to those of Multiple Regresssion
#'     Analysis. \emph{Journal of the Japanese Society for Quality Control,
#'     42}(2), 103-115. (In Japanese)
#'
#'   Kawada, H., & Nagata, Y. (2015). An application of a generalized inverse
#'     regression estimator to Taguchi's T-Method. \emph{Total Quality Science,
#'     1}(1), 12-21.
#'
#' @seealso \code{\link{general_T}},
#'            \code{\link{generates_transformation_functions_T1}}, and
#'            \code{\link{forecasting.Ta}}
#'
#' @examples
#' model_Ta <- Ta(sample_data = stackloss[-c(2, 12, 19), ],
#'                subtracts_V_e = TRUE,
#'                includes_transformed_data = TRUE)
#'
#' (model_Ta$M_hat)
#'
#' @export
Ta <- function(sample_data,
               subtracts_V_e = TRUE,
               includes_transformed_data = FALSE) {

  model_Ta <- general_T(unit_space_data = sample_data,
                        signal_space_data = sample_data,
                        generates_transform_functions =
                                          generates_transformation_functions_T1,
                        subtracts_V_e = subtracts_V_e,
                        includes_transformed_data = includes_transformed_data)

  class(model_Ta) <- "Ta"

  return(model_Ta)

}

#' Forecasting method for the Ta method
#'
#' \code{forecasting.Ta} (via \code{\link{forecasting}}) estimates the dependent
#'   values based on the Ta model.
#'
#' @param model Object of class "Ta" generated by \code{\link{Ta}} or
#'                \code{\link{generates_model}}(..., method = "Ta").
#' @param newdata Matrix with n rows (samples) and p columns (variables). The
#'                  Data to be estimated. All data should be continuous values
#'                  and should not have missing values.
#' @param includes_transformed_newdata If \code{TRUE}, then the transformed data
#'                                       for \code{newdata} are included in a
#'                                       return object.
#'
#' @return A list containing the following components is returned.
#'
#'  \item{M_hat}{Vector with length n. The estimated values of the dependent
#'                variable after the data transformation.}
#'  \item{y_hat}{Vector with length n. The estimated values after the inverse
#'                transformation from \code{M_hat}.}
#'  \item{model}{Object of class "Ta" passed by \code{model}.}
#'  \item{n}{The number of samples for \code{newdata}.}
#'  \item{q}{The number of variables after the data transformation. q equals p.}
#'  \item{X}{If \code{includes_transformed_newdata} is \code{TRUE}, then the
#'            transformed data for \code{newdata} are included.}
#'
#' @references
#'   Inou, A., Nagata, Y., Horita, K., & Mori, A. (2012). Prediciton Accuracies
#'     of Improved Taguchi's T Methods Compared to those of Multiple Regresssion
#'     Analysis. \emph{Journal of the Japanese Society for Quality Control,
#'     42}(2), 103-115. (In Japanese)
#'
#'   Kawada, H., & Nagata, Y. (2015). An application of a generalized inverse
#'     regression estimator to Taguchi's T-Method. \emph{Total Quality Science,
#'     1}(1), 12-21.
#'
#' @seealso \code{\link{general_forecasting.T}} and \code{\link{Ta}}
#'
#' @examples
#' model_Ta <- Ta(sample_data = stackloss[-c(2, 12, 19), ],
#'                subtracts_V_e = TRUE,
#'                includes_transformed_data = TRUE)
#'
#' forecasting_Ta <- forecasting(model = model_Ta,
#'                               newdata = stackloss[c(2, 12, 19), -4],
#'                               includes_transformed_newdata = TRUE)
#'
#' (forecasting_Ta$y_hat) # Estimated values
#' (stackloss[c(2, 12, 19), 4]) # True values
#'
#' @export
forecasting.Ta <- function(model,
                           newdata,
                           includes_transformed_newdata = FALSE) {

  if (!inherits(model, "Ta")) {
    warning("calling forecasting.Ta(<fake-Ta-object>) ...")
  }

  general_forecasting.T(model = model,
                        newdata = newdata,
                        includes_transformed_newdata =
                                                   includes_transformed_newdata)

}


R/wrapper.R
#' Wrapper function to generate a unit space for a family of
#'   Mahalanobis-Taguchi (MT) methods
#'
#' \code{generates_unit_space} generates a unit space for a family of
#'   Mahalanobis-Taguchi (MT) methods. The unit space of \code{\link{MT}} method,
#'   \code{\link{MTA}} method or \code{\link{RT}} method can be generated by
#'   passing a method name (character) into a parameter \code{method}.
#'
#' @param unit_space_data Matrix with n rows (samples) and p columns (variables).
#'                          Data to generate the unit space. All data should be
#'                          continuous values and should not have missing values.
#' @param method Character to designate a method. Currently, "MT", "MTA", and
#'                 "RT" are available.
#' @param includes_transformed_data If \code{TRUE}, then the transformed data
#'                                    are included in a return object.
#' @param ... Passed to \code{\link[base]{solve}} for computing the inverse of
#'              the correlation matrix in \code{\link{MT}} and \code{\link{RT}}
#'              method.
#'
#' @return A returned object depends on the selected method. See \code{\link{MT}},
#'           \code{\link{MTA}} or \code{\link{RT}}.
#'
#' @seealso \code{\link{MT}}, \code{\link{MTA}}, \code{\link{RT}}, and
#'            \code{\link[base]{solve}}
#'
#' @examples
#' # 40 data for versicolor in the iris dataset
#' iris_versicolor <- iris[61:100, -5]
#'
#' # 10 data for each kind (setosa, versicolor, virginica) in the iris dataset
#' iris_test <- iris[c(1:10, 51:60, 101:111), -5]
#'
#' # MT method
#' unit_space_MT <- generates_unit_space(unit_space_data = iris_versicolor,
#'                                       method = "MT")
#'
#' diagnosis_MT <- diagnosis(unit_space = unit_space_MT,
#'                           newdata = iris_test,
#'                           threshold = 4)
#'
#' (diagnosis_MT$distance)
#' (diagnosis_MT$le_threshold)
#'
#' # MTA method
#' unit_space_MTA <- generates_unit_space(unit_space_data = iris_versicolor,
#'                                        method = "MTA")
#'
#' diagnosis_MTA <- diagnosis(unit_space = unit_space_MTA,
#'                            newdata = iris_test,
#'                            threshold = 0.5)
#'
#' (diagnosis_MTA$distance)
#' (diagnosis_MTA$le_threshold)
#'
#' # RT method
#' unit_space_RT <- generates_unit_space(unit_space_data = iris_versicolor,
#'                                       method = "RT")
#'
#' diagnosis_RT <- diagnosis(unit_space = unit_space_RT,
#'                           newdata = iris_test,
#'                           threshold = 0.2)
#'
#' (diagnosis_RT$distance)
#' (diagnosis_RT$le_threshold)
#'
#' @export
generates_unit_space <- function(unit_space_data,
                                method = c("MT", "MTA", "RT"),
                                includes_transformed_data = FALSE,
                                ...) {

  method <- match.arg(method)

  if (method == "MT") {
    MT(unit_space_data, includes_transformed_data, ...)
  } else if (method == "MTA") {
    MTA(unit_space_data, includes_transformed_data)
  } else if (method == "RT") {
    RT(unit_space_data, includes_transformed_data, ...)
  } else {
  }

}

#' Wrapper function to generate a model for a family of Taguchi (T) methods
#'
#' \code{generates_model} generates a model for a family of Taguchi (MT) methods.
#'   The model of \code{\link{T1}} method, \code{\link{Ta}} method or the
#'   \code{\link{Tb}} method can be generated by passing a method name
#'   (character) into a parameter \code{method}.
#'
#' @param unit_space_data Used only for the T1 method. Matrix with n rows
#'                          (samples) and (p + 1) columns (variables). The 1 ~ p
#'                          th columns are independent variables and the (p + 1)
#'                          th column is a dependent variable. Underlying data
#'                          to obtain a representative point for the
#'                          normalization of \code{signal_space_data}. All data
#'                          should be continuous values and should not have
#'                          missing values.
#' @param signal_space_data Used only for the T1 method. Matrix with m rows
#'                            (samples) and (p + 1) columns (variables). The
#'                            1 ~ p th columns are independent variables and the
#'                            (p + 1) th column is a dependent variable.
#'                            Underlying data to generate a prediction
#'                            expression. All data should be continuous values
#'                            and should not have missing values.
#' @param sample_data Used for the Ta and the Tb methods. Matrix with n rows
#'                      (samples) and (p + 1) columns (variables). The 1 ~ p th
#'                      columns are independent variables and the (p + 1) th
#'                      column is a dependent variable. All data should be
#'                      continuous values and should not have missing values.
#' @param method Character to designate a method. Currently, "MT", "MTA", and
#'                 "RT" are available.
#' @param subtracts_V_e If \code{TRUE}, then the error variance is subtracted in
#'                        the numerator when calculating \code{eta_hat}.
#' @param includes_transformed_data If \code{TRUE}, then the transformed data
#'                                    are included in a return object.
#'
#' @return A returned object depends on the selected method. See \code{\link{T1}},
#'           \code{\link{Ta}} or \code{\link{Tb}}.
#'
#' @seealso \code{\link{T1}}, \code{\link{Ta}}, \code{\link{Tb}}
#'
#' @examples
#' # The value of the dependent variable of the following samples mediates
#' # in the stackloss dataset.
#' stackloss_center <- stackloss[c(9, 10, 11, 20, 21), ]
#'
#' # The following samples are data other than the unit space data and the test
#' # data.
#' stackloss_signal <- stackloss[-c(2, 9, 10, 11, 12, 19, 20, 21), ]
#'
#' # The following test samples are chosen casually.
#' stackloss_test <- stackloss[c(2, 12, 19), -4]
#'
#' # T1 method
#' model_T1 <- generates_model(unit_space_data = stackloss_center,
#'                             signal_space_data = stackloss_signal,
#'                             method = "T1",
#'                             subtracts_V_e = TRUE)
#'
#' forecasting_T1 <- forecasting(model = model_T1,
#'                               newdata = stackloss_test)
#'
#' (forecasting_T1$y_hat)
#'
#' # Ta method
#' model_Ta <- generates_model(sample_data =
#'                                    rbind(stackloss_center, stackloss_signal),
#'                             method = "Ta",
#'                             subtracts_V_e = TRUE)
#'
#' forecasting_Ta <- forecasting(model = model_Ta,
#'                               newdata = stackloss_test)
#'
#' (forecasting_Ta$y_hat)
#'
#' # Tb method
#' model_Tb <- generates_model(sample_data =
#'                                    rbind(stackloss_center, stackloss_signal),
#'                             method = "Tb",
#'                             subtracts_V_e = TRUE)
#'
#' forecasting_Tb <- forecasting(model = model_Tb,
#'                               newdata = stackloss_test)
#'
#' (forecasting_Tb$y_hat)
#'
#' @export
generates_model <- function(unit_space_data,
                            signal_space_data,
                            sample_data,
                            method = c("T1", "Ta", "Tb"),
                            subtracts_V_e = TRUE,
                            includes_transformed_data = FALSE) {

  method <- match.arg(method)

  if (method == "T1") {

    if (!missing(sample_data)) {
      warning("Parameter \"sample_data\" is not used in T1 method.")
    }

    T1(unit_space_data = unit_space_data,
       signal_space_data = signal_space_data,
       subtracts_V_e = subtracts_V_e,
       includes_transformed_data = includes_transformed_data)

  } else if (method == "Ta") {

    if (!missing(unit_space_data)) {
      warning("Parameter \"unit_space_data\" is not used in Ta method.")
    }

    if (!missing(signal_space_data)) {
      warning("Parameter \"signal_space_data\" is not used in Ta method.")
    }

    Ta(sample_data = sample_data,
       subtracts_V_e = subtracts_V_e,
       includes_transformed_data = includes_transformed_data)

  } else if (method == "Tb") {

    if (!missing(unit_space_data)) {
      warning("Parameter \"unit_space_data\" is not used in Tb method.")
    }

    if (!missing(signal_space_data)) {
      warning("Parameter \"signal_space_data\" is not used in Tb method.")
    }

    Tb(sample_data = sample_data,
       subtracts_V_e = subtracts_V_e,
       includes_transformed_data = includes_transformed_data)

  } else {
  }

}
